src/components/TreeGrid.ts

```diff
@@ -46,13 +46,43 @@ const wrapContent =
     state: 'closed',
   });
 
+const hierarchyMapToRows = <T extends RowContents>(hierarchyMap: Map<number, T[]>, domainOptionRoot: T): Row<T>[] => {
+  const traverseHierarchy = (parent: T, depth: number, previousRows: Row<T>[]): Row<T>[] => {
+    // does parent have children?
+    const children = hierarchyMap.get(parent.id) || [];
+    const parentRow: Row<T> = {
+      contents: parent,
+      depth,
+      isFetched: children.length > 0,
+      state: children.length > 0 ? 'open' : 'closed',
+    };
+    // recursively traverse hierarchy of all children
+    const childRows: Row<T>[] = children.flatMap((child) => traverseHierarchy(child, depth + 1, []));
+    return [...previousRows, parentRow, ...childRows];
+  };
+
+  const rows: Row<T>[] = [];
+  const depth = 0;
+  // get all children of domain option root
+  const domainOptionRootChildren = hierarchyMap.get(domainOptionRoot.id) || [];
+
+  domainOptionRootChildren.forEach((child) => {
+    rows.push(...traverseHierarchy(child, depth, []));
+  });
+
+  return rows;
+};
+
 type TreeGridProps<T extends RowContents> = {
   /** the columns to display */
   readonly columns: Column<T>[];
   /** the initial rows to display */
-  readonly initialRows: T[];
+  readonly initialHierarchy: Map<number, T[]>;
   /** Given a row, return its children. This is only called if row.hasChildren is true. */
   readonly getChildren: (row: T) => Promise<T[]>;
+  /** Given the domain option root, create a hierarchy */
+  readonly domainOptionRoot: T;
```

Comment:
I'd like to keep snapshot builder out of `TreeGrid` if possible. In terms of `TreeGrid`, this is the tree's root.
```suggestion
  /** The root of the tree's hierarchy. */
  readonly root: T;
```

Reply:
I've removed :)